Add error codes to API

Kwaidan00 opened the pull request:
I want to merge this change because it adds error codes for mutation errors.

It closes #4119.

<!-- Please mention all relevant issue numbers. -->

### Screenshots

<!-- If your changes affect the UI, providing "before" and "after" screenshots will
greatly reduce the amount of work needed to review your work. -->

### Pull Request Checklist

<!-- Please keep this section. It will make maintainer's life easier. -->

1. [ ] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

NyanKiyoshi commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,177 @@
+]
+
+
+class AccountErrorCode(Enum):
```
I think it would be better to put them to the concerned packages, i.e. `AccountErrorCode` into `graphql.account`.

Not sure where. The current convention for enums is to put them in the `__init__.py` but I'm not sure if that's the best place, especially as we don't always expect to have other stuff than imports inside this file.

NyanKiyoshi replied:
cc. @maarcingebala

Kwaidan00 replied:
I would be glad to hear your suggestion. I do not insist with that solution. :)

maarcingebala replied:
Fixed, now they sit next to models.

maarcingebala commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,182 @@
+    MAX_LESS_THAN_MIN = "max_less_than_min"
+
+
+ERROR_CODE_UNKNOWN = "unknown"
```
I'm not sure what I think about returning "unknown" code. First of all of shouldn't have such errors. Secondly, maybe instead we should just return `null` in that case?

Kwaidan00 replied:
We should not, but something must be returned if some bad error code was used. 
For simplicity I've changed it to `null`.

maarcingebala commented on saleor/graphql/menu/mutations.py:
```diff
@@ -254,9 +280,12 @@ def clean_operation(operation: _MenuMoveOperation):
+                                "Cannot assign a node as child of "
+                                "one of its descendants."
+                            ),
+                            code=MenuErrorCode.CANNOT_ASSIGN_NOTE,
```
Looks like a typo, should be `CANNOT_ASSIGN_NODE`? But I think this one and `ASSIGN_MENU_ITEM_TO_ITSELF` are in fact the same error which tells you that this particular assignment is not allowed.

Kwaidan00 replied:
Yes, it was a typo.
I've unified those two error codes.

maarcingebala commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,182 @@
+    DELETE_NON_STAFF_USER = "delete_non_staff_user"
+    DELETE_OWN_ACCOUNT = "delete_own_account"
+    DELETE_STAFF_ACCOUNT = "delete_staff_account"
+    DELETE_STUFF_ACCOUNT = "delete_stuff_account"
```
`DELETE_STAFF_ACCOUNT ` vs `DELETE_STUFF_ACCOUNT `

maarcingebala commented on saleor/graphql/order/mutations/draft_orders.py:
```diff
@@ -302,7 +310,12 @@ def perform_mutation(cls, _root, info, **data):
+                    {
+                        "quantity": ValidationError(
+                            "Ensure this value is greater than or equal to 1.",
+                            code=OrderErrorCode.QUANTITY_LESS_THAN_ONE,
```
We have at least these codes that are similar: `POSITIVE_NUMBER_REQUIRED`, QUANTITY_LESS_THAN_ONE`. I think that instead, we should have a few generic ones such as `LESS_THAN`, `LESS_THAN_OR_EQUAL_TO` and the same for "greater" checks.

maarcingebala commented on saleor/graphql/product/mutations/attributes.py:
```diff
@@ -160,15 +171,29 @@ def clean_attribute(cls, instance, cleaned_input):
+                {
+                    "slug": ValidationError(
+                        "The attribute's slug cannot be blank.",
+                        code=AttributeErrorCode.ATTRIBUTE_SLUG_BLANK,
```
This is simply a "value required" type of error.

Kwaidan00 replied:
I've changed it to Django's `blank` error code.

maarcingebala commented:
I did the first review and I think we have to look through the codes and think whether we can unify/generalize some of them. Another question is whether we need to start exposing them all at once. Maybe a better idea would be to have dedicated enums for each group e.g. `CheckoutErrorCodes`.

Kwaidan00 commented on saleor/core/utils/promo_code.py:
```diff
@@ -4,19 +4,29 @@
+            message = {
+                "promo_code": ValidationError(
+                    "Promo code already exists.",
+                    code=CheckoutErrorCode.PROMO_CODE_ALREADY_EXISTS,
```
I don't think there is any benefit. In some context it could be more readable, but I think simplicity is better.
I've unified it.

Kwaidan00 commented on saleor/graphql/account/mutations/account.py:
```diff
@@ -120,7 +124,13 @@ def perform_mutation(cls, _root, info, **data):
+            raise ValidationError(
+                {
+                    "token": ValidationError(
+                        INVALID_TOKEN, code=AccountErrorCode.INVALID_USER_TOKEN
```
I've added generic invalid token code.

Kwaidan00 commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,182 @@
+from enum import Enum
+
+DJANGO_VALIDATORS_ERROR_CODES = [
```
I would be glad if we were able to do it prettier. Unfortunately, Django doesn't store any list of errors or something. Every code that throws `ValidationError` uses string literals as the value of code parameter.
On the other hand, we can not drop those codes, because there are returned by model and forms validation.

django-queries commented:
Here is the report for 43625e6f8eb5f6fe80ab829ab7061a08c18db103 (mirumee/saleor @ add-error-codes-in-api)
Base comparison is 1ac9086f38296b2986c2a884e780b20ff5be3355.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  add billing address to checkout     	         34	         34	             20
  add shipping to checkout            	          7	          7	              0
  checkout payment charge             	         14	         14	              0
  complete checkout                   	          6	          6	              0
  create checkout                     	         48	         48	             24

# api.benchmark homepage
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  retrieve main menu                  	          5	          5	              0
  retrieve product list               	          4	          4	              0
  retrieve secondary menu             	          5	          5	              0
  retrieve shop                       	          2	          2	              0

# api.benchmark product
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  product details                     	         13	         13	              3

# api.benchmark variant
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  retrieve variant list               	         15	         15	              8
```

</p>
</details>

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,169 @@
+    ORDER_NO_PAYMENT = "order_no_payment"
+    ORDER_NO_PRODUCTS = "order_without_products"
+    ORDER_NO_SHIPPING_ADDRESS = "order_no_shipping_address"
+    QUANTITY_LESS_THAN_ONE = "quantity_less_than_one"
```
This is not true. What we require is a non-zero quantity. We might as well allow decimals like 0.2 in the future. I propose that we name it `ZERO_QUANTITY` and fix the error message.

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,169 @@
+    ORDER_NO_PRODUCTS = "order_without_products"
+    ORDER_NO_SHIPPING_ADDRESS = "order_no_shipping_address"
+    QUANTITY_LESS_THAN_ONE = "quantity_less_than_one"
+    REFUND_MANUAL_PAYMENT = "refund_manual_payment"
```
Does it need to be specific like that? How about `CANNOT_REFUND` to match the first two?

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,169 @@
+    EDIT_NON_DRAFT_ORDER = "edit_non_draft_order"
+    FULFILL_ORDER_LINE = "fulfill_order_line"
+    ORDER_INVALID_SHIPPING_METHOD = "order_invalid_shipping_method"
+    ORDER_NO_PAYMENT = "order_no_payment"
```
`PAYMENT_MISSING`? `ORDER_NO_PAYMENT` sounds like a music group from Tokyo.

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,169 @@
+    CANNOT_CANCEL_ORDER = "cannot_cancel_order"
+    CAPTURE_INACTIVE_PAYMENT = "capture_inactive_payment"
+    DELETE_LINE_NON_DRAFT_ORDER = "delete_line_non_draft_order"
+    DELETE_NON_DRAFT_ORDER = "delete_non_draft_order"
```
`CANNOT_DELETE`?

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,169 @@
+    CANNOT_CANCEL_FULFILLMENT = "cannot_cancel_fulfillment"
+    CANNOT_CANCEL_ORDER = "cannot_cancel_order"
+    CAPTURE_INACTIVE_PAYMENT = "capture_inactive_payment"
+    DELETE_LINE_NON_DRAFT_ORDER = "delete_line_non_draft_order"
```
`CANNOT_DELETE`? Do we need a separate error code for lines and orders?

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,169 @@
+    INVALID_COUNTRY = "invalid_country"
+    INVALID_PASSWORD = "invalid_password"
+    INVALID_PHONE_NUMBER = "invalid_phone_number"
+    NOT_USERS_ADDRESS = "not_users_address"
```
Why is this useful? Do we ever need the UI to know that an address ID is correct but belongs to a different user?

patrys commented:
Can we plan a meeting where we go through all of these codes and determine whether they are correct or even needed?

Kwaidan00 commented:
@patrys @maarcingebala I've created typed versions of `Error` type and separated enums. Please take a look.

maarcingebala commented on saleor/graphql/product/mutations/digital_contents.py:
```diff
@@ -165,7 +170,10 @@ def clean_input(cls, info, data):
-                raise ValidationError(msg.format(*missing_fields))
+                raise ValidationError(
+                    msg.format(*missing_fields),
+                    code=ProductErrorCode.MISSING_CONFIGURATION_FIELDS,
```
`REQUIRED` would work here I guess.

maarcingebala commented on saleor/graphql/schema.graphql:
```diff
@@ -2573,9 +2858,38 @@ input ProductCreateInput {
+  ATTRIBUTE_NON_ASSIGNABLE
+  ATTRIBUTE_VALUES_NOT_UNIQUE
+  GRAPHQL_ERROR
+  INCORRECT_VALUE
```
Can't see where this is used. Maybe we could replace it with `INVALID`?

maarcingebala commented on saleor/graphql/account/mutations/base.py:
```diff
@@ -39,7 +41,29 @@ def can_edit_address(user, address):
+    )
+
+    @classmethod
+    def handle_typed_errors(cls, errors: list, **extra):
```
For every mixing like this, we need to override the entire method and the implementation is the same in every case. What if we kept this implementation in the BaseMutation only and in the mixins just provided a constant like `ERROR_TYPE_CLASS=AccountError` etc that we could use in `BaseMutation`? I would experiment with something like this to avoid code duplication.

Kwaidan00 replied:
I wanted to do it that way, but I've stuck with that at some point. I'll refactor it.

Kwaidan00 replied:
I've added the second field to the `cls`, for typed error field name. I think it is more readable than e..g searching class attribute with some special name.

patrys commented on saleor/graphql/schema.graphql:
```diff
@@ -5,27 +5,63 @@ schema {
+  INVALID_PASSWORD
+  INVALID_PHONE_NUMBER
+  INVALID_STOREFRONT_URL
+  INVALID_TOKEN
```
Not sure about the password: should we separate the cases of a password being invalid and a password being too weak? I'm thinking about the registration flow, where the user might want to know whether to add more random characters or to remove a character that is not allowed.

Kwaidan00 replied:
I've kept invalid password error code and I've merged all the others.

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,225 @@
+    ATTRIBUTE_BAD_VALUE = "attribute_bad_value"
+    ATTRIBUTE_CANNOT_BE_ASSIGNED = "attribute_cannot_be_assigned"
+    ATTRIBUTE_DISABLED_VARIANTS = "attribute_disabled_variants"
+    ATTRIBUTE_NON_ASSIGNABLE = "attribute_non_assignable"
```
How is this different from "cannot be assigned"?

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,225 @@
+    GRAPHQL_ERROR = "graphql_error"
+    INCORRECT_VALUE = "incorrect_value"
+    INVALID = "invalid"
+    MISSING_CONFIGURATION_FIELDS = "missing_configuration_fields"
```
Shouldn't those fields have a `REQUIRED` error?

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,225 @@
+    MISSING_CONFIGURATION_FIELDS = "missing_configuration_fields"
+    MISSING_VALUE = "missing_value"
+    NOT_PRODUCTS_IMAGE = "not_products_image"
+    NULL = "null"
```
How is `NULL` different from `REQUIRED`?

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,225 @@
+    OBJECT_DOES_NOT_EXIST = "object_does_not_exist"
+    REQUIRED = "required"
+    UNIQUE = "unique"
+    VALUE_ERROR = "value_error"
```
Is this not the same as `INVALID`?

patrys commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,225 @@
+    INVALID = "invalid"
+    NULL = "null"
+    OBJECT_ALREADY_EXISTS = "object_already_exists"
+    OBJECT_DOES_NOT_EXIST = "object_does_not_exist"
```
Those two look strangely generic.

Kwaidan00 replied:
Is it bad?

patrys replied:
When would we return these? Not just for this enum, for all enums. There's usually a better way to indicate what went wrong like returning `CUSTOMER_ALREADY_EXISTS` or returning `ALREADY_EXISTS` accompanied by a proper field name.

Kwaidan00 replied:
I think all necessary information is given with that error code by attaching it to the correct field. However, I've replaced `OBJECT_ALREADY_EXISTS` with more specific error codes, as you suggested.

Refactoring of `OBJECT_DOES_NOT_EXIST ` is more complicated. I think e.g. `get_node_or_error` should be rewritten and should create field-specific codes - or we have to remain some generic error code for that. I think it could be done via separated PR.

patrys replied:
Yeah, I guess `get_node_or_error` should start accepting error codes.

maarcingebala replied:
But is the current implementation acceptable or is this something that we definitely want to change before merging?

maarcingebala commented on saleor/graphql/account/mutations/base.py:
```diff
@@ -39,7 +41,21 @@ def can_edit_address(user, address):
+        graphene.NonNull(AccountError),
+        description="List of errors that occurred executing the mutation.",
+    )
+    ERROR_TYPE_CLASS = AccountError
```
Maybe let's just add one empty line above, to visually separate fields declaration from those mixin settings.

Kwaidan00 replied:
I've added a line :)

maarcingebala commented:
@patrys @NyanKiyoshi If you could take a look - I've made few changes: moved error codes definitions to separate files and replaced:
- `OBJECT_DOES_NOT_EXIST` with `NOT_FOUND`
- `OBJECT_ALREADY_EXISTS` with `ALREADY_EXISTS` (plus unified a few other "already exists" errors)

korycins commented on saleor/core/exceptions.py:
```diff
@@ -1,4 +1,8 @@
+from ..checkout.error_codes import CheckoutErrorCode
```
Core module should not take any dependencies from other Saleor's module.

maarcingebala replied:
Ideally yes; in this case, the exception is either used in `product` / `checkout` or `order` modules. Maybe instead of `core`, it should be defined inside one of them (not sure which one though). We will anyway end up importing it between those modules. @korycins @k-brk  If you have better idea then we can move it, if not I would leave it here.

Kwaidan00 replied:
I've created a ticket for that: #4744 
We can decide it later.

korycins commented on saleor/core/utils/promo_code.py:
```diff
@@ -9,14 +9,19 @@
         if message is None:
-            message = {"promo_code": "Promo code is invalid"}
+            message = {
+                "promo_code": ValidationError("Promo code is invalid", code="invalid")
```
Shouldn't we use `ErrorCode` class for `code` here?

Kwaidan00 replied:
I've added an import from `GiftCardError` here; PR for changing that: #4744

korycins commented on saleor/core/utils/promo_code.py:
```diff
@@ -9,14 +9,19 @@
 
 class PromoCodeAlreadyExists(ValidationError):
     def __init__(self, message=None, **kwargs):
+        code = kwargs.get("code", "promo_code_already_exists")
```
Shouldn't we use `ErrorCode` class here?

Kwaidan00 replied:
As above.

korycins commented on saleor/core/utils/url.py:
```diff
@@ -15,10 +17,14 @@ def validate_storefront_url(url):
     except ValueError as error:
-        raise ValidationError({"redirectUrl": str(error)})
+        raise ValidationError(
+            {"redirectUrl": str(error)}, code=AccountErrorCode.INVALID
```
Same as above. We shouldn't import any saleor's modules in core. This is a perfect wat to get more circular imports.
If we need to import `AccountErrorCode` it means that this class or `validate_storefront_url` is declared in wrong place.

Kwaidan00 replied:
It will stay as it is for now; I've created a ticket for that: #4744

korycins commented on saleor/core/weight.py:
```diff
@@ -98,7 +98,8 @@ def validate(self, weight):
                 raise forms.ValidationError(
-                    "Invalid unit: %r (expected %r)." % (weight.unit, unit)
+                    "Invalid unit: %r (expected %r)." % (weight.unit, unit),
+                    code="invalid",
```
Shouldn't we use `ErrorCode` class here?

Kwaidan00 replied:
I cannot see any way to avoid it for now; I've created a ticket for that, as we discussed: #4745

korycins commented on saleor/graphql/account/mutations/base.py:
```diff
@@ -137,19 +177,26 @@ def perform_mutation(cls, _root, info, **data):
             password_validation.validate_password(new_password, user)
         except ValidationError as error:
-            raise ValidationError({"new_password": error.messages})
+            raise ValidationError({"new_password": error})
```
Shouldn't return error code here?

Kwaidan00 replied:
`error` here is a `ValidationError` instance and it should already have error code.

korycins commented on saleor/graphql/giftcard/mutations.py:
```diff
@@ -8,12 +8,29 @@
+        description="List of errors that occurred executing the mutation.",
+    )
+
+    ERROR_TYPE_CLASS = GiftcardError
```
`GiftcardError` -> `GiftCardError`

korycins commented on saleor/graphql/giftcard/mutations.py:
```diff
@@ -8,12 +8,29 @@
 from .types import GiftCard
 
 
+class GiftcardErrorMixin:
```
`GiftcardError` -> `GiftCardError`

korycins commented on saleor/graphql/menu/mutations.py:
```diff
@@ -7,16 +7,33 @@
 from .types import Menu, MenuItem, MenuItemMoveInput
 
 
+class MenuErrorMixin:
```
Do we really want to declare and maintain base mutation for each module?
It seems scary for me. 
Why don't we use `__init_subclass_with_meta__` in BaseMutation to dynamically declare expected fields for the given class?.
```
class MenuUpdate(ModelMutation):
    class Meta:
        ERROR_TYPE_CLASS = MenuError
        ERROR_TYPE_FIELD = "menu_errors"
        error_type = MenuError
```
`__init_subclass_with_meta__` could use this fields to generate error structure. This will add some magic to base mutations but even though I think that this is better than declaring new base mutation per module.

Kwaidan00 replied:
The last commit brings that implementation. ;)

maarcingebala commented on saleor/account/error_codes.py:
```diff
@@ -0,0 +1,22 @@
+from enum import Enum
+
+
+class AccountErrorCode(Enum):
```
If some of them are common to ALL enums, then we could have one base enum for that.

Kwaidan00 replied:
@maarcingebala We discussed with @korycins and we decided to change enums to classes. The problem is that you cannot extend enum with fields (https://docs.python.org/3/library/enum.html#restricted-enum-subclassing). We don't use any enums special logic, so class-based approach would resolve that problem.

patrys replied:
This sound counterintuitive. Subclassing means it's hard to see all members without executing code and custom classes cannot be used for typing.

Kwaidan00 replied:
Ok, I won't change it.

NyanKiyoshi commented on saleor/graphql/product/mutations/attributes.py:
```diff
@@ -239,7 +263,13 @@ def clean_remove_values(cls, cleaned_input, instance):
+                raise ValidationError(
+                    {
+                        "remove_values": ValidationError(
+                            msg, code=ProductErrorCode.ATTRIBUTE_BAD_VALUE
```
```suggestion
                            msg, code=ProductErrorCode.INVALID
```

NyanKiyoshi commented on saleor/graphql/product/mutations/attributes.py:
```diff
@@ -387,7 +426,12 @@ def perform_mutation(
+                {
+                    "operations": ValidationError(
+                        "Variants are disabled in this product type.",
+                        code=ProductErrorCode.ATTRIBUTE_DISABLED_VARIANTS,
```
```suggestion
                        code=ProductErrorCode.ATTRIBUTE_VARIANTS_DISABLED,
```

NyanKiyoshi commented on saleor/graphql/product/mutations/products.py:
```diff
@@ -1157,7 +1229,12 @@ def perform_mutation(cls, _root, info, image_id, variant_id):
+                {
+                    "image_id": ValidationError(
+                        "Image is not assigned to this variant.",
+                        code=ProductErrorCode.NOT_PRODUCTS_IMAGE,
```
```suggestion
                        code=ProductErrorCode.NOT_FOUND,
```

Kwaidan00 replied:
@NyanKiyoshi are you sure it's the same? There is a different between non-existing image and existing image that not belongs to that product.

patrys commented on saleor/graphql/account/mutations/staff.py:
```diff
@@ -210,11 +212,21 @@ def clean_is_active(cls, is_active, instance, user):
+                    {
+                        "is_active": ValidationError(
+                            "Cannot deactivate superuser's account.",
+                            code=AccountErrorCode.DEACTIVATE_SUPERUSER_ACCOUNT,
```
🤔 Wouldn't that mean that most if not all error codes would start with CANNOT? I think I'd prefer to avoid negatives and use errors like `ACCOUNT_PROTECTED`. Ideally, the error code could be used to finish the following sentence: "the operation failed because...".

Kwaidan00 commented on saleor/graphql/order/utils.py:
```diff
@@ -16,15 +25,25 @@ def validate_shipping_method(order):
+            {
+                "shipping": ValidationError(
+                    "Shipping method is not valid for chosen shipping address",
+                    code=OrderErrorCode.ORDER_INVALID_SHIPPING_METHOD,
```
I would prefer to have a separated error code here. It's an important part of checkout process and I think it should have that code.

NyanKiyoshi replied:
Fair enough. Maybe we could say `SHIPPING_METHOD_NOT_APPLICABLE` or something like this? As it is not really an invalid shipping method but it's simply no applying to the active checkout.

Kwaidan00 replied:
You're right; I've changed it to `SHIPPING_METHOD_NOT_APPLICABLE`.

Kwaidan00 commented on saleor/graphql/core/utils/__init__.py:
```diff
@@ -27,15 +27,24 @@ def str_to_enum(name):
+            {
+                field: ValidationError(
+                    "Couldn't resolve to a node: %s" % global_id,
+                    code="object_does_not_exist",
```
Yes, it's hardcoded - moreover, with bad value. 
But I think it must be hardcoded - we cannot extend some base enum, so I cannot decide which `SthErrorCode` should I import to that module.
The alternative way is passing correct enum when that function is called - but I'm not sure it's good idea.

Kwaidan00 replied:
@NyanKiyoshi I've created a ticket for later refactoring: #4745

Kwaidan00 commented on saleor/graphql/core/utils/error_codes.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def get_error_code_from_error(error):
+    """Return valid error code."""
```
I've added it.

Kwaidan00 commented:
@maarcingebala @korycins @NyanKiyoshi I've changed error code mechanism to use `Meta` class instead of inheritance, for some of your comments I've created separated issues, I've rebased branch and added error codes for new parts of the codebase.
